server/src/main/java/org/opensearch/identity/IdentityService.java

```diff
@@ -27,12 +30,12 @@ public class IdentityService {
     private final Settings settings;
     private final IdentityPlugin identityPlugin;
 
-    public IdentityService(final Settings settings, final List<IdentityPlugin> identityPlugins) {
+    public IdentityService(final Settings settings, final ThreadPool threadPool, final List<IdentityPlugin> identityPlugins) {
```

Comment:
I think there is potential with the code changes here, and I'd like to see us close the loop on the scenario before this is merged.  We need to settle on design the plugin experience *first*.

Taking this class from `ml-commons` repo, it jumps into the system context, how should it look instead? 
```java
public class SearchTaskTransportAction extends HandledTransportAction<SearchRequest, SearchResponse> {
    private Client client;

    @Inject
    public SearchTaskTransportAction(TransportService transportService, ActionFilters actionFilters, Client client) {
        super(MLTaskSearchAction.NAME, transportService, actionFilters, SearchRequest::new);
        this.client = client;
    }

    @Override
    protected void doExecute(Task task, SearchRequest request, ActionListener<SearchResponse> actionListener) {
        try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
            final ActionListener<SearchResponse> wrappedListener = ActionListener
                .wrap(actionListener::onResponse, e -> wrapListenerToHandleSearchIndexNotFound(e, actionListener));
            client.search(request, ActionListener.runBefore(wrappedListener, () -> context.restore()));
        } catch (Exception e) {
            log.error(e.getMessage(), e);
            actionListener.onFailure(e);
        }
    }
}
```
_[Ref](
https://github.com/opensearch-project/ml-commons/blob/2a33c652d79b733121e3340bf9edc1f7993b7590/plugin/src/main/java/org/opensearch/ml/action/tasks/SearchTaskTransportAction.java#L25C1-L45C2)_

Reading the user case for this method SearchTaskTransportAction is searching for ML models in a private index.  So the transport action comes in with user context and they need permissions to the private index.

I think we could get away with building a slightly different version of `HandledTransportAction` that had an additional method `T runAsPlugin(Function<PluginNodeClient, T> operation)`, slimming this down:

```java
    ...
    @Override
    protected void doExecute(Task task, SearchRequest request, ActionListener<SearchResponse> actionListener) {
       final ActionListener<SearchResponse> wrappedListener = ActionListener
          .wrap(actionListener::onResponse, e -> wrapListenerToHandleSearchIndexNotFound(e, actionListener));
       this.pluginClient(pluginNodeClient -> pluginNodeClient.search(request, wrappedListener));
    }
    ...
```

So if ^ is the solution then lets build up what we need to support that scenario.  Maybe it can't be as clean as I'm making it out to be[1], but this has a clean value proposition to plugin teams.  Lets make this PR focus on that clear value, the great news is I feel like we've got a lot of the critical building blocks.

- [1] For example maybe we need to pass something additional into the injected constructor, and that needs to be piped into where getRestHandlers(...) is declared.

Reply:
@peternied See the sample plugin in the Companion PR for how adoption of this change will/can look like: https://github.com/opensearch-project/security/pull/4665/files#diff-0609767672c3297dd70f9882e0407c0e0ed517eda81eba53d0aa443b0e0597b9

Reply:
While I think there are improvements that can be done here this looks like a solid step forward - with some additional steps incremental improvements to prevent plugin developers from making security missteps.  Resolving this thread.